app/components/request/configs.js

```diff
@@ -16,63 +16,106 @@ export default Component.extend(CanTriggerBuild, TriggerBuild, {
   classNames: ['request-configs'],
   classNameBindings: ['status'],
 
-  requestConfig: service(),
+  preview: service('request-config'),
   features: service(),
-
-  status: 'closed',
-  closed: equal('status', 'closed'),
-  customizing: equal('status', 'customize'),
-  previewing: equal('status', 'preview'),
-  replacing: equal('mergeMode', 'replace'),
   showNewConfigView: reads('features.showNewConfigView'),
 
   repo: reads('request.repo'),
-  rawConfigs: or('requestConfig.rawConfigs', 'request.uniqRawConfigs'),
-  messages: reads('requestConfig.messages'),
-  loading: reads('requestConfig.loading'),
+  repoId: reads('repo.id'),
+  rawConfigs: or('preview.rawConfigs', 'request.uniqRawConfigs'),
+  messages: reads('preview.messages'),
+  loading: reads('preview.loading'),
 
   refType: 'sha',
-  sha: truncated('originalSha', 10),
+  sha: reads('originalSha'),
   branch: reads('originalBranch'),
   message: reads('request.commit.message'),
   config: reads('request.apiConfig.config'),
   mergeMode: reads('originalMergeMode'), // TODO store and serve merge mode for api request configs
   defaultMergeMode: 'deep_merge_append',
 
-  originalSha: or('requestSha', 'repoDefaultBranchLastCommitSha'),
+  originalSha: truncated('requestOrDefaultBranchSha', 10),
   originalBranch: or('requestBranch', 'repoDefaultBranch'),
   originalMergeMode: or('request.mergeMode', 'defaultMergeMode'),
   requestBranch: reads('request.branchName'),
   requestSha: reads('request.commit.sha'),
+  requestOrDefaultBranchSha: or('requestSha', 'repoDefaultBranchLastCommitSha'),
   repoDefaultBranch: reads('repo.defaultBranch.name'),
   repoDefaultBranchLastCommitSha: reads('repo.defaultBranch.lastBuild.commit.sha'),
 
-  onTriggerBuild(e) {
+  ref: computed('refType', 'branch', 'sha', function () {
+    return this.get(this.refType);
+  }),
+
+  didInsertElement() {
+    this.load();
+  },
+
+  onTrigger(e) {
     e.toElement.blur();
-    if (this.status == 'closed') {
+    if (this.closed) {
       this.set('status', 'open');
-    } else if (['open', 'customize', 'preview'].includes(this.status)) {
+    } else {
       this.submit();
     }
   },
 
   onCustomize() {
-    if (this.status == 'customize') {
+    if (this.customizing) {
       this.set('status', 'open');
       this.reset();
     } else {
       this.set('status', 'customize');
     }
   },
 
+  onPreview() {
+    if (!this.previewing) {
+      this.set('status', 'preview');
+      if (!this.loaded) {
+        this.load();
+      }
+    } else if (this.customized) {
+      this.set('status', 'customize');
+    } else {
+      this.set('status', 'open');
+    }
+  },
+
   onCancel() {
     this.set('status', 'closed');
     this.reset();
   },
 
+  load(debounce) {
+    if (this.customizing || this.previewing) {
+      this.preview.loadConfigs.perform(this.repoId, this.data, debounce);
+      this.set('loaded', true);
```

Comment:
oh, most probably, yes. i'll try that.

done: https://github.com/travis-ci/travis-web/commit/9881f3578ef33be42209b6de6700ea663adf227d

i had to move it to the service because otherwise it would raise when restarted. but this looks better to me anyway.

Reply:
@svenfuchs thank you for the fix! Although, I'd like to suggest a little bit different approach: https://github.com/travis-ci/travis-web/commit/9881f3578ef33be42209b6de6700ea663adf227d#r37998445 Please take a look when you got a chance and let me know what you think.

Reply:
thanks! i've commented over there with a question ...

Reply:
i have added that question to a new section "to be discussed/resolved" on https://github.com/travis-ci/travis-web/pull/2399 in order to keep track of this